File: 1557-check-if-a-string-contains-all-binary-codes-of-size-k/check-if-a-string-contains-all-binary-codes-of-size-k.java
class Solution {
    public boolean hasAllCodes(String s, int k) {
        int totalNeeded = 1 << k; // This is 2^k
        boolean[] seen = new boolean[totalNeeded];
        int count = 0;
        
        // A mask to keep only the last k bits (e.g., if k=3, mask is 111 in binary)
        int mask = totalNeeded - 1;
        int currentRollingHash = 0;

        for (int i = 0; i < s.length(); i++) {
            // Push the new bit into the rolling hash
            currentRollingHash = ((currentRollingHash << 1) | (s.charAt(i) - '0')) & mask;

            // Once we have seen at least k bits, start tracking
            if (i >= k - 1) {
                if (!seen[currentRollingHash]) {
                    seen[currentRollingHash] = true;
                    count++;
                    // Optimization: If we found all codes, stop early
                    if (count == totalNeeded) return true;
                }
            }
        }

        return false;
    }
}

File: 1557-check-if-a-string-contains-all-binary-codes-of-size-k/check-if-a-string-contains-all-binary-codes-of-size-k.py
class Solution:
    def hasAllCodes(self, s: str, k: int) -> bool:
        # We need to find 2^k unique substrings
        total_needed = 1 << k
        seen = set()

        # Slide the window across the string
        # s[i : i + k] takes a slice of length k
        for i in range(len(s) - k + 1):
            substring = s[i : i + k]
            if substring not in seen:
                seen.add(substring)
                # Optimization: Exit early if we found everything
                if len(seen) == total_needed:
                    return True
                    
        return False